lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -800,75 +826,72 @@ private void tryWait(long timeout) {
 		}
 	}
 
-	public void run() {
-		Thread thisThread = Thread.currentThread();
-		while(!state.terminal) {
-
-			/*
-			 * Until we've reached a terminal state we:
-			 * - get a state change;
-			 * - enact that change
-			 */
-			StateIndication stateChange = null;
-
-			/* Hold the lock until we obtain a state change */
-			synchronized(this) {
-				/* if we're initialising, then tell the starting thread that
-				 * we're ready to receive events */
-				if(state.state == ConnectionState.initialized) {
-					synchronized(thisThread) {
-						thisThread.notify();
+	class CMThread extends Thread {
+
+		private boolean exiting = false;
+		private void setExiting() {
+			exiting = true;
+		}
+
+		public void run() {
+			ConnectionManager cm = ConnectionManager.this;
+			while(!exiting) {
+				/*
+				 * Until we're commited to exit we:
+				 * - get a state change;
+				 * - enact that change
+				 */
+				StateIndication stateChange = null;
+
+				/* Hold the lock until we obtain a state change */
+				synchronized(cm) {
+					/* if we're initialising, then tell the starting thread that
+					 * we're ready to receive events */
+					if (state.state == ConnectionState.initialized) {
+						synchronized(this) {
+							notify();
+						}
 					}
-				}
 
-				while(!state.terminal && stateChange == null) {
-					/* wait for a state change event or for expiry of the current state */
-					tryWait(state.timeout);
+					while(!exiting && stateChange == null) {
```

Comment:
Aren't we already in a `!exiting` context given https://github.com/ably/ably-java/pull/404/files?utf8=%E2%9C%93&diff=split#diff-060685a947a7032665de7ac9d33bb6ffR838 ? Do we need this double check?

Reply:
All we know is that we were `!exiting` when we entered the loop at the top. Since then we will have waited for state changes, so it's no longer expected to be true.